rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+    InteractiveContext and basically does nothing / returns empty
+    ExecutionResult on .run().
+
+    ```
+    class DummyInteractiveContext(InteractiveContext):
+    def run(self,
+      component: base_component.BaseComponent,
+      enable_cache: bool = True):
+    return None
+    ```
+
+    1.  This should cover the case where the class definition is aliased.
+
+        ```
+        aliased_class = interactive_context.InteractiveContext
+        sess = aliased_class()
+        ```
+
+    1.  This should cover subsequent aliases of InteractiveContext instances.
+
+        ```
+        a = InteractiveContext()
+        b = a
+        c = InteractiveContext()
+        d = c
+        ```
+
+Cons:
+
+*   DummyInteractiveContext is now present/clutters the production pipeline code
+    (it's a no-op so mainly affects readability, not execution).
+*   Down the line, converting back to a notebook (replacing
+    DummyInteractiveContext with InteractiveContext) could be fragile.
+
+##### Option 2: Ensure InteractiveContext only runs in notebook context. **[recommended]**
+
+*   If InteractiveContext is run outside of a notebook context, just log a
+    warning and return.
+*   Bi-directional import to notebook from pipeline would "just work".
+
+Cons:
+
+*   InteractiveContext is still present in the production pipeline as a no-op /
+    affects readability.
+*   Puts the burden on user to scrub out calls to InteractiveContext.
+
+##### Option 3: Mark lines/cells to be skipped during export.
+
+Add custom magic to mark lines/cells as skip_for_export, can also be used by the
+user to skip scratch work in cells.
+
+Example line magic:
+
+```
+%skip_for_export sess = InteractiveContext()
+...
+%skip_for_export sess.run(example_gen)
+```
+
+Example cell magic:
+
+```
+%%skip_for_export
+# Cell contains scratch work that doesn't need to be exported.
+...
+```
+
+Cons:
+
+*   Puts burden on the user to filter out the InteractiveContext objects. User
+    may forget to mark some usages of InteractiveContext, meaning
+    InteractiveContext instances can get leaked to the final pipeline.
+
+##### Option 4: Delete the lines containing InteractiveContext variables.
+
+Cons:
+
+*   Not robust to duplicate references.
+*   We can find the references to InteractiveContext by either keeping track of
+    them weakly within the class on __init__, or we can use gc module to
+    dynamically find the references. But then finding and deleting all
+    associated lines with each instance seems hard.
+    *   What if user makes a helper function and passes in a context variable?
+        (not likely, but possible)
+
+Note each of these options only filters the InteractiveContext usage in the
+exported python script, and does not prevent the user from adding it back
+afterwards.
+
+#### Export notebook contents to pipeline
+
+1.  Present the user with a Beam pipeline export template cell. Airflow/Kubeflow
```

Comment:
What do you mean by "Present the user with..."-- is that in documentation? How is it presented?